File: fishola-mobile/src/pojos/UserProfile.ts
/*-
 * #%L
 * Fishola :: Mobile
 * %%
 * Copyright (C) 2019 - 2021 INRAE - UMR CARRTEL
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * #L%
 */

export default class UserProfile {
  static currentUser?: UserProfile;

  id: string;
  lastName?: string;
  gender?: string;
  birthYear?: number;
  offlineMarker: boolean = false;

  constructor(
    public firstName: string,
    public email: string,
    public pseudo: string,
    public initials: string,
    public sampleBaseId: string,
    public acceptsMailNotifications: boolean,
    public acceptsShareTrips: boolean,
    public lastNewsSeenDate: Date
  ) {}

  static fullName(p: UserProfile) {
    let result = p.firstName;
    if (p.lastName) {
      result += " " + p.lastName;
    }
    return result;
  }

  static fromJson(input: any) {
    const result = new UserProfile(
      input.firstName,
      input.email,
      input.pseudo,
      input.initials,
      input.sampleBaseId,
      input.acceptsMailNotifications,
      input.acceptsShareTrips,
      input.lastNewsSeenDate,
    );
    result.lastName = input.lastName;
    result.gender = input.gender;
    result.birthYear = input.birthYear;
    return result;
  }

  static getCurrent(): UserProfile {
    return this.currentUser!;
  }

  static setCurrent(newProfile: UserProfile) {
    this.currentUser = newProfile;
  }

  static unsetCurrent() {
    delete this.currentUser;
  }
}

File: fishola-backend/src/main/java/fr/inrae/fishola/rest/security/UserProfile.java
package fr.inrae.fishola.rest.security;

/*-
 * #%L
 * Fishola :: Backend
 * %%
 * Copyright (C) 2019 - 2021 INRAE - UMR CARRTEL
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * #L%
 */

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.google.common.base.Joiner;
import com.google.common.base.Splitter;
import fr.inrae.fishola.ImmutableObject;
import fr.inrae.fishola.entities.enums.Gender;
import org.immutables.value.Value;

import java.time.LocalDateTime;
import java.util.LinkedList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

@ImmutableObject
@JsonSerialize(as = ImmutableUserProfile.class)
@JsonDeserialize(as = ImmutableUserProfile.class)
public interface UserProfile {

    UUID id();

    String firstName();

    Optional<String> lastName();

    String email();

    Optional<Gender> gender();

    Optional<Integer> birthYear();

    String sampleBaseId();

    Boolean acceptsMailNotifications();

    Boolean acceptsShareTrips();

    LocalDateTime lastNewsSeenDate();

    @Value.Derived
    default String initials() {
        List<Character> initials = new LinkedList<>(splitToInitials(firstName()));
        lastName().map(this::splitToInitials)
                .ifPresent(initials::addAll);
        String joined = Joiner.on("").join(initials).toUpperCase();
        return joined;
    }

    default List<Character> splitToInitials(String input) {
        List<String> parts = Splitter.on(Pattern.compile("[ -]"))
                .trimResults()
                .omitEmptyStrings()
                .splitToList(input.trim());
        List<Character> result = parts.stream()
                .map(part -> part.charAt(0))
                .toList();
        return result;
    }

}
